File: 서예진A/20230518/H-index.cpp
#include <vector>
#include <algorithm>
using namespace std;


int solution(vector<int> citations) {
    
    int answer = 0;
    
    sort(citations.begin(), citations.end(), greater<>()); //내림차순 정렬
    

    for(int i=0; i<citations.size(); i++){
        if(citations[i] >= i+1){
            answer = i+1;
        }  
    }

    return answer;
}

// #include <string>
// #include <vector>
// #include <algorithm>

// using namespace std;

// int solution(vector<int> citations) {
//     int answer = 0;
//     int size = citations.size();
//     int max = 0;
    
//     // 인용횟수 가장 큰 값 구하기 
//     for(int i=0;i<size;i++){
//         max = max < citations[i] ? citations[i] : max;
//     }
        
//     int cnt = 0; //h번 이상 인용된 논문의 갯수 count
    
//     //가장 많이 인용된 횟수 -1 해가면서 그 이상 인용된 논문 수보다 그 횟수가 작아지는 경우 찾아내기
//     for(int i=max; i>-1; i--)
//     {
//         for(int j=0;j<size;j++)
//         {
//             if(citations[j]>=i) cnt++;
//         }
        
//         if(cnt>=i) 
//         {
//             answer = i;
//             break;
//         }
//     }
    
//     return answer;
// }

File: 서예진B/20230518/H-Index.py
def solution(citations):
    answer = 0
    n = len(citations)
    # h를 논문 개수로 초기화
    h = n
    
    # h를 줄여가며 h의 값이 적절한지 체크
    # O(n^2)인데 n이 1000이라서 가능
    while h >= 0:
        cited = 0
        less_cited = 0
        # citations 배열을 순회하며 논문 개수 카운팅
        for i in range(n):
            if citations[i] >= h:
                # h번 이상 인용된 논문 수 + 1
                cited += 1
            elif citations[i] <= h:
                # h번 이하로 인용된 논문 수 + 1
                less_cited += 1
                
        # h번 인용된 논문 수가 h개 이상, 나머지(n-cited)가 h번 이하로 인용
        if cited >= h and less_cited == n - cited:
            return h
    
        h -= 1
    '''
    엄청난 사람의 코드 ..
    
    citations.sort()
    for i in range(n):
        if citations[i] >= n - i:
            answer += 1
    '''
    return answer